### PR TITLE
Task panel UI improvements

### PR DESCRIPTION
**Pull request by dde6khkg**
# @ mention of reviewers
@ckcollab @Logan-Ruf 

# A brief description of the purpose of the changes contained in this PR.
This PR adds the ability to share tasks with other users. It also adds two columns to the task management table: benchmarks associated with the task, and users the task is shared with.

# Issues this PR resolves
Addresses #416 

# Known issues to be addressed in a separate PR
#626 #627 

# A checklist for hand testing
- [x] Upload a competition
- [x] Go to the Resources page, on the Tasks tab
- [x] Click a task and click share, select some users and click share.
- [ ] Login as a user that was shared to.
- [ ] Check that the shared task is accessible.

# Checklist
- [x] Code review by me 
- [x] Hand tested by me 
- [x] I'm proud of my work
- [ ] Code review by reviewer
- [ ] Hand tested by reviewer
- [ ] Ready to merge

### CODE REVIEW

File: src/static/riot/tasks/management.tag [hunk 1]
```diff
@@ -28,6 +30,8 @@
         <tbody>
         <tr each="{ task in tasks }" onclick="{show_detail_modal.bind(this, task)}" class="task-row">
             <td>{ task.name }</td>
+            <td class="benchmark-row">{ task.competitions.join(', ') }</td>
+            <td>{ task.shared_with.join(', ') }</td>
```
Comment (Logan-Ruf) [inline src/static/riot/tasks/management.tag:34]: When you change from using method fields to nested serializer you will have to change this to something like.
```suggestion
            <td class="benchmark-row">{ task.competitions.title }, { task.competitions.id }</td>
            <td>{ task.shared_with.name }</td>
```
You could even add a link to the user's profile if you think it fits.

**Review by Logan-Ruf — CHANGES_REQUESTED** [review]

File: src/apps/api/views/tasks.py [hunk 2]
```diff
@@ -37,17 +40,20 @@ def get_queryset(self):
...
             # passing submission
             task_validate_qs = Submission.objects.filter(
                 md5__in=OuterRef("solutions__md5"),
                 status=Submission.FINISHED,
-                phase__in=OuterRef("phases")
+                # TODO: This line causes our query to take ~10 seconds. Is this important?
+                # phase__in=OuterRef("phases")
             )
             # We have to grab something from task_validate_qs here, so i grab pk
             qs = qs.annotate(validated=Subquery(task_validate_qs.values('pk')[:1]))
```
[the remark this answers is not in the document]
Comment (ckcollab) [reply src/apps/api/views/tasks.py:59]: we made this way faster actually just a few moments ago, that is correct though. 

I believe the rationale was to keep things simple, making this data "live" is relatively easy vs making it into a property we have to sync whenever submissions are deleted/such. we want a valid submission that exists

File: src/apps/api/serializers/tasks.py [hunk 3]
```diff
@@ -113,6 +116,8 @@ def get_validated(self, task):
 class TaskListSerializer(serializers.ModelSerializer):
     solutions = SolutionListSerializer(many=True, required=False, read_only=True)
     value = serializers.CharField(source='key', required=False)
+    competitions = serializers.SerializerMethodField()
+    shared_with = serializers.SerializerMethodField()
```
[the remark this answers is not in the document]
Comment (dde6khkg) [reply src/apps/api/serializers/tasks.py:120]: The method fields are used to reduce queries. I'm getting everything I need in the viewset in 1 or two queries and passing what I need as context. I think doing it that way will cause queries to scale with tasks right?

**Review by Logan-Ruf — CHANGES_REQUESTED** [review]
Share modal doesn't clear after you share then open another task.
image

**Comment by Logan-Ruf** [comment]
Tested sharing tasks and making sure those tasks are shown to the user that it was shared too.

**Pull request merged**